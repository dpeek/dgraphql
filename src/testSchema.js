/* @flow */

import fs from 'fs'
import path from 'path'

import { buildSchema } from './schema'
import { Client } from './client'
import type { ClientConfig } from './client'

export default async function initSchema (
  graphql: string,
  dgraph: string,
  config: ClientConfig
) {
  const client = new Client(config, {})
  let dgraphPath = path.resolve(__dirname, '__tests__', dgraph)
  await client.fetchQuery(fs.readFileSync(dgraphPath).toString())

  const graphqlPath = path.resolve(__dirname, '__tests__', graphql)
  return buildSchema(fs.readFileSync(graphqlPath).toString(), config)
}
