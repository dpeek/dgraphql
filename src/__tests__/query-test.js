import { graphql } from 'graphql'
import testSchema from '../testSchema'

let schema

beforeAll(async () => {
  schema = await testSchema('test.graphql', 'test.dgraph', {})
})

test('queries node field', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('queries node fields', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim",
      active: true
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      name
      active
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('queries type name', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      __typename
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult.data.person.__typename).toEqual('Person')
})

test('queries aliased field', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      test: name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('queries nested field', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim",
      partner: {
        name: "Bob"
      }
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      name
      partner {
        name
      }
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

describe('querying edges', () => {
  let time = 0
  beforeAll(() => {
    time = String(new Date().getTime() - 1495660000000)
    const create = `mutation {
      aaron: createPerson(input: {
        name: "Aaron Whitman",
        active: true,
        age:20,
        time: ${time}
      }) {
        person { id }
      }
      bobby: createPerson(input: {
        name: "Bobby Whitman",
        active: false,
        age: 16,
        time: ${time}
      }) {
        person { id }
      }
      catherine: createPerson(input: {
        name: "Catherine Harrison",
        active: true,
        age: 19,
        time: ${time}
      }) {
        person { id }
      }
    }`
    return graphql(schema, create)
  })
  test('order edges by string ascending', async () => {
    const query = `query {
      people(
        order: name_asc,
        filter: {
          time_eq: ${time}
        }
      ) {
        name
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('order edges by string descending', async () => {
    const query = `query {
      people(
        order: name_desc,
        filter: {
          time_eq: ${time}
        }
      ) {
        name
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by string with all terms', async () => {
    const query = `query {
      people(
        order: name_asc,
        filter: {
          time_eq: ${time},
          name_allofterms: "Aaron Whitman"
        }
      ) {
        name
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by string with any of terms', async () => {
    const query = `query {
      people(
        order: name_asc,
        filter: {
          time_eq: ${time},
          name_anyofterms: "Aaron Catherine"
        }
      ) {
        name
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by string equal to', async () => {
    const query = `query {
      people(
        order: name_asc,
        filter: {
          time_eq: ${time},
          name_eq: "Catherine Harrison"
        }
      ) {
        name
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by int equal to', async () => {
    const query = `query {
      people(
        order: age_asc,
        filter: {
          time_eq: ${time},
          age_eq: 19
        }
      ) {
        name
        age
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by int less than', async () => {
    const query = `query {
      people(
        order: age_asc,
        filter: {
          time_eq: ${time},
          age_lt: 19
        }
      ) {
        name
        age
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by int less than or equal to', async () => {
    const query = `query {
      people(
        order: age_asc,
        filter: {
          time_eq: ${time},
          age_le: 19
        }
      ) {
        name
        age
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by int greater than or equal to', async () => {
    const query = `query {
      people(
        order: age_asc,
        filter: {
          time_eq: ${time},
          age_gt: 19
        }
      ) {
        name
        age
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
  test('filters by int greater than or equal to', async () => {
    const query = `query {
      people(
        order: age_asc,
        filter: {
          time_eq: ${time},
          age_ge: 19
        }
      ) {
        name
        age
      }
    }`
    const queryResult = await graphql(schema, query)
    expect(queryResult).toMatchSnapshot()
  })
})
