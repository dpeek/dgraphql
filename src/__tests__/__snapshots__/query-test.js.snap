// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters by boolean equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "employed": true,
        "name": "Amy Winter",
      },
      Object {
        "employed": true,
        "name": "David Peek",
      },
    ],
  },
}
`;

exports[`filters by float equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "height": 1.7,
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`filters by float greater than 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "height": 1.8,
        "name": "Matthew Peek",
      },
      Object {
        "height": 1.99,
        "name": "David Peek",
      },
    ],
  },
}
`;

exports[`filters by float greater than or equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "height": 1.7,
        "name": "Amy Winter",
      },
      Object {
        "height": 1.8,
        "name": "Matthew Peek",
      },
      Object {
        "height": 1.99,
        "name": "David Peek",
      },
    ],
  },
}
`;

exports[`filters by float less than 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "height": 1.12,
        "name": "Olivia Peek",
      },
      Object {
        "height": 1.6,
        "name": "Linda Peek",
      },
    ],
  },
}
`;

exports[`filters by float less than or equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "height": 1.12,
        "name": "Olivia Peek",
      },
      Object {
        "height": 1.6,
        "name": "Linda Peek",
      },
      Object {
        "height": 1.7,
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`filters by int equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "age": 37,
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`filters by int greater than 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "age": 37,
        "name": "Amy Winter",
      },
      Object {
        "age": 65,
        "name": "Linda Peek",
      },
      Object {
        "age": 70,
        "name": "Matthew Peek",
      },
    ],
  },
}
`;

exports[`filters by int greater than or equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "age": 37,
        "name": "Amy Winter",
      },
      Object {
        "age": 65,
        "name": "Linda Peek",
      },
      Object {
        "age": 70,
        "name": "Matthew Peek",
      },
    ],
  },
}
`;

exports[`filters by int less than 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "age": 3,
        "name": "Olivia Peek",
      },
      Object {
        "age": 35,
        "name": "David Peek",
      },
      Object {
        "age": 37,
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`filters by int less than or equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "age": 3,
        "name": "Olivia Peek",
      },
      Object {
        "age": 35,
        "name": "David Peek",
      },
      Object {
        "age": 37,
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`filters by string equal to 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "Olivia Peek",
      },
    ],
  },
}
`;

exports[`filters by string with all terms 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "David Peek",
      },
    ],
  },
}
`;

exports[`filters by string with any of terms 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "David Peek",
      },
      Object {
        "name": "Olivia Peek",
      },
    ],
  },
}
`;

exports[`order edges by string ascending 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "Amy Winter",
      },
      Object {
        "name": "David Peek",
      },
      Object {
        "name": "Linda Peek",
      },
      Object {
        "name": "Matthew Peek",
      },
      Object {
        "name": "Olivia Peek",
      },
    ],
  },
}
`;

exports[`order edges by string descending 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "name": "Olivia Peek",
      },
      Object {
        "name": "Matthew Peek",
      },
      Object {
        "name": "Linda Peek",
      },
      Object {
        "name": "David Peek",
      },
      Object {
        "name": "Amy Winter",
      },
    ],
  },
}
`;

exports[`queries aliased field 1`] = `
Object {
  "data": Object {
    "person": Object {
      "test": "David Peek",
    },
  },
}
`;

exports[`queries nested field 1`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "David Peek",
      "partner": Object {
        "name": "Amy Winter",
      },
    },
  },
}
`;

exports[`queries node field 1`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "David Peek",
    },
  },
}
`;

exports[`queries node fields 1`] = `
Object {
  "data": Object {
    "person": Object {
      "employed": true,
      "name": "David Peek",
    },
  },
}
`;
